packages/backend/src/model/BlockRange.ts

```diff
@@ -0,0 +1,76 @@
+import assert from 'assert'
+import { range } from 'lodash'
+
+import { BlockNumber } from '../peripherals/ethereum/types'
+
+export class BlockRange {
+  /** number of the earliest block in range */
+  public readonly from: BlockNumber
+  /** number of the latest block in range */
+  public readonly to: BlockNumber
+  private readonly hashes: ReadonlyMap<BlockNumber, string>
+
+  constructor(
+    /** a list of blocks in order from oldest to newest */
+    blocks: readonly Block[]
+  )
+  constructor(
+    /** parent block range */
+    range: BlockRange,
+    /** new left boundary */
+    from: BlockNumber,
+    /** new right boundary */
+    to: BlockNumber
+  )
+
+  constructor(...args: BlockRangeConstructorArgs) {
+    if (args.length === 3) {
+      const [range, from, to] = args
+      this.to = to
+      this.from = from
+      this.hashes = range.hashes
+      return
+    }
+
+    const [blocks] = args
+    assert(blocks.length > 0, 'BlockRange must have at least one block')
+
+    this.from = blocks[0].number
```

Comment:
This should be the smallest, we aren't guaranteed that the blocks are given in-order